File: ParteJava/ProyectoPro3/ProyectoPro3Persistencia/src/main/java/pe/edu/pucp/proyectopro3/dao/notificaciones/NotificacionDAOImpl.java
package pe.edu.pucp.proyectopro3.dao.notificaciones;

import pe.edu.pucp.proyectopro3.dao.DefaultBaseDAO;
import pe.edu.pucp.proyectopro3.modelo.notificaciones.Notificacion;

import java.sql.*;

public class NotificacionDAOImpl extends DefaultBaseDAO<Notificacion> implements NotificacionDAO {

    @Override
    protected PreparedStatement comandoCrear(Connection conn, Notificacion modelo) throws SQLException {
        String sql = "{call sp_InsertNotificacion(?, ?, ?, ?, ?, ?)}";
        CallableStatement cmd = conn.prepareCall(sql);

        setCamposNotificacion(cmd, modelo, 1);
        cmd.registerOutParameter(6, Types.INTEGER);

        return cmd;
    }

    @Override
    protected PreparedStatement comandoActualizar(Connection conn, Notificacion modelo) throws SQLException {
        String sql = "{call sp_UpdateNotificacion(?, ?, ?, ?, ?, ?)}";
        CallableStatement cmd = conn.prepareCall(sql);

        cmd.setInt(1, modelo.getIdNotificacion());
        setCamposNotificacion(cmd, modelo, 2);

        return cmd;
    }

    @Override
    protected PreparedStatement comandoEliminar(Connection conn, Integer id) throws SQLException {
        String sql = "{call sp_DeleteNotificacion(?)}";
        CallableStatement cmd = conn.prepareCall(sql);
        cmd.setInt(1, id);
        return cmd;
    }

    @Override
    protected PreparedStatement comandoLeer(Connection conn, Integer id) throws SQLException {
        String sql = "{call sp_ListNotificacionById(?)}";
        CallableStatement cmd = conn.prepareCall(sql);
        cmd.setInt(1, id);
        return cmd;
    }

    @Override
    protected PreparedStatement comandoLeerTodos(Connection conn) throws SQLException {
        String sql = "{call spListNotificaciones()}";
        return conn.prepareCall(sql);
    }

    @Override
    protected Integer extraerIdDesdeCallable(CallableStatement cmd) throws SQLException {
        return cmd.getInt(6);
    }

    @Override
    protected Integer extraerIdDesdeGeneratedKeys(ResultSet rs) throws SQLException {
        return rs.getInt(1);
    }

    @Override
    protected Notificacion mapearModelo(ResultSet rs) throws SQLException {
        Notificacion notificacion = new Notificacion();

        notificacion.setIdNotificacion(rs.getInt("id_notificacion"));
        notificacion.setMensaje(rs.getString("mensaje"));
        notificacion.setTipoNotificacion(rs.getString("tipo_notificacion"));
        notificacion.setFechaEnvio(rs.getTimestamp("fecha_envio"));
        notificacion.setLeido("Y".equalsIgnoreCase(rs.getString("leido")));
        notificacion.setIdUsuario(rs.getInt("id_usuario"));

        return notificacion;
    }

    private void setCamposNotificacion(CallableStatement cmd, Notificacion modelo, int inicio) throws SQLException {
        cmd.setString(inicio, modelo.getMensaje());
        cmd.setString(inicio + 1, obtenerTipoNotificacion(modelo));
        cmd.setTimestamp(inicio + 2, new java.sql.Timestamp(modelo.getFechaEnvio().getTime()));
        cmd.setString(inicio + 3, modelo.isLeido() ? "Y" : "N");
        cmd.setInt(inicio + 4, modelo.getIdUsuario());
    }

    private String obtenerTipoNotificacion(Notificacion modelo) {
        if (modelo.getTipoNotificacion() != null && !modelo.getTipoNotificacion().isBlank()) {
            return modelo.getTipoNotificacion();
        }
        if (modelo.getTipoEvento() != null) {
            return modelo.getTipoEvento().name();
        }
        return "GENERAL";
    }
}

File: ParteC#/ProyectoProgramacion3/ProyectoProgramacion3Persistencia/DAO/notificaciones/NotificacionDaoImpl.cs
using System.Data;
using System.Data.Common;
using ProyectoProgramacion3Model.Model.notificaciones;

namespace ProyectoProgramacion3Persistencia.DAO.notificaciones;

public class NotificacionDaoImpl : DefaultBaseDao<Notificacion>, INotificacionDao
{
    protected override DbCommand ComandoCrear(DbConnection conn, Notificacion modelo)
    {
        var cmd = CrearStoredProcedure(conn, "sp_InsertNotificacion");
        SetCamposNotificacion(cmd, modelo);
        CrearParametroSalida(cmd, "_id_generado", DbType.Int32);
        return cmd;
    }

    protected override DbCommand ComandoActualizar(DbConnection conn, Notificacion modelo)
    {
        var cmd = CrearStoredProcedure(conn, "sp_UpdateNotificacion");
        CrearParametro(cmd, "p_id_notificacion", modelo.idNotificacion);
        SetCamposNotificacion(cmd, modelo);
        return cmd;
    }

    protected override DbCommand ComandoEliminar(DbConnection conn, int id)
    {
        var cmd = CrearStoredProcedure(conn, "sp_DeleteNotificacion");
        CrearParametro(cmd, "p_id_notificacion", id);
        return cmd;
    }

    protected override DbCommand ComandoLeer(DbConnection conn, int id)
    {
        var cmd = CrearStoredProcedure(conn, "sp_ListNotificacionById");
        CrearParametro(cmd, "p_id_notificacion", id);
        return cmd;
    }

    protected override DbCommand ComandoLeerTodos(DbConnection conn)
    {
        return CrearStoredProcedure(conn, "spListNotificaciones");
    }

    protected override Notificacion MapearModelo(DbDataReader reader)
    {
        return new Notificacion
        {
            idNotificacion = LeerEntero(reader, "id_notificacion"),
            mensaje = LeerTexto(reader, "mensaje"),
            tipoNotificacion = LeerTexto(reader, "tipo_notificacion"),
            fechaEnvio = LeerDateOnly(reader, "fecha_envio"),
            leido = LeerBoolYN(reader, "leido"),
            idUsuario = LeerEntero(reader, "id_usuario")
        };
    }

    private void SetCamposNotificacion(DbCommand cmd, Notificacion modelo)
    {
        CrearParametro(cmd, "p_mensaje", modelo.mensaje);
        CrearParametro(cmd, "p_tipo_notificacion", ObtenerTipoNotificacion(modelo));
        CrearParametro(cmd, "p_fecha_envio", DateOnlyParam(modelo.fechaEnvio), DbType.DateTime);
        CrearParametro(cmd, "p_leido", modelo.leido ? "Y" : "N");
        CrearParametro(cmd, "p_id_usuario", modelo.idUsuario);
    }

    private static string ObtenerTipoNotificacion(Notificacion modelo)
    {
        if (!string.IsNullOrWhiteSpace(modelo.tipoNotificacion))
        {
            return modelo.tipoNotificacion;
        }

        return modelo.tipoEvento.ToString() ?? "GENERAL";
    }
}
